pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -895,24 +897,27 @@ private void pendingBatchReceiveTask(Timeout timeout) throws Exception {
     protected void triggerListener() {
         // Trigger the notification on the message listener in a separate thread to avoid blocking the networking
         // thread while the message processing happens
-        Message<T> msg;
-        do {
-            try {
-                msg = internalReceive(0, TimeUnit.MILLISECONDS);
+        try {
+            // Control executor to call MessageListener one by one.
+            if (executorQueueSize.get() < 1) {
```

Comment:
Do we still need the executorQueueSize since we don't have a loop for triggering the listener?

Reply:
LGTM, the incoming messages will trigger the listener and add the message to the thread pool, so the executorQueueSize is required.